src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,57 @@
 module Juvix.Backends.ArithmeticCircuit.Compilation.Types where
 
-import qualified Data.Graph.Inductive as G
+import qualified Juvix.Core.ErasedAnn as CoreErased
 import Juvix.Library
+import qualified Juvix.Backends.ArithmeticCircuit.Parameterisation as Parameterisation
+import qualified Circuit.Expr as Expr
+import qualified Circuit
 
-type Circuit = G.Gr NodeType WireType
+data PrimVal f
+  = Constant f
+  | FEInteger f
+  | BinOp BinOp (PrimVal f) (PrimVal f)
+  | Op Op (PrimVal f)
 
-data NodeType
-  = -- Addition gate.
-    Add
-  | -- Multiplication gate.
-    Mul
-  | -- Input value (by vector index).
-    InputValue Natural
-  | -- Output value (by vector index).
-    OutputValue Natural
-  deriving (Show, Eq, Generic)
+data BinOp
+  = Add
+  | Mul
+  | Sub
+  | Exp
+  | Eq
 
-type WireType = ()
+data Op = Neg
+
+type Term = CoreErased.AnnTerm () (PrimVal Parameterisation.F)
+type Type = CoreErased.Type () (PrimVal Parameterisation.F)
+
+data CompilationError
+  = NotYetImplemented
+  | SomethingWentWrongSorry
+
+data Memory x where
+  Mem :: Parameterisation.FieldElement x => [x] -> Memory x
+
+eval :: Mem x -> Term -> (Mem x, Expr.Expr Circuit.Wire Parameterisation.F a)
+eval mem CoreErased.AnnTerm{ CoreErased.term = CoreErased.Term (CoreErased.Prim term) } = (mem, translateTerm term)
+eval mem CoreErased.AnnTerm{ CoreErased.term = CoreErased.Term (CoreErased.Var var) } = (mem, Expr.EVar v)
+eval (Mem ms) CoreErased.AnnTerm{ CoreErased.term = CoreErased.Term CoreErased.LamM{ CoreErased.arguments = arguments }}
```

Comment:
Should we deal with this? I think we will run into this case when compiling `λx.λw. w^2 == x`, for example. Maybe you want to specifically match out the first two arguments, and inline everything else?